api/transport/invoker.go

```diff
@@ -0,0 +1,133 @@
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"time"
+
+	"go.uber.org/yarpc/yarpcerrors"
+	"go.uber.org/zap"
+)
+
+// UnaryInvokeRequest encapsulates minimum arguments to invoke a unary handler
+type UnaryInvokeRequest struct {
+	Context        context.Context
+	StartTime      time.Time
+	Request        *Request
+	ResponseWriter ResponseWriter
+	Handler        UnaryHandler
+	Options        *InvokerOptions
+}
+
+// OnewayInvokeRequest encapsulates minimum arguments to invoke a unary handler
+type OnewayInvokeRequest struct {
+	Context context.Context
+	Request *Request
+	Handler OnewayHandler
+	Options *InvokerOptions
+}
+
+// StreamInvokeRequest encapsulates minimum arguments to invoke a unary handler
+type StreamInvokeRequest struct {
+	Stream  *ServerStream
+	Handler StreamHandler
+	Options *InvokerOptions
+}
+
+// InvokeUnaryHandler calls the handler h, recovering panics and timeout errors,
+// converting them to yarpc errors. All other errors are passed through.
+func InvokeUnaryHandler(
+	i UnaryInvokeRequest,
+) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = handlePanic(Unary, i.Options, r, i.Request.ToRequestMeta())
+		}
+	}()
+
+	err = i.Handler.Handle(i.Context, i.Request, i.ResponseWriter)
+
+	// The handler stopped work on context deadline.
+	if err == context.DeadlineExceeded && err == i.Context.Err() {
+		deadline, _ := i.Context.Deadline()
+		err = yarpcerrors.Newf(
+			yarpcerrors.CodeDeadlineExceeded,
+			"call to procedure %q of service %q from caller %q timed out after %v",
+			i.Request.Procedure, i.Request.Service, i.Request.Caller, deadline.Sub(i.StartTime))
+	}
+	return err
+}
+
+// InvokeOnewayHandler calls the oneway handler, recovering from panics as
+// errors
+func InvokeOnewayHandler(
+	i OnewayInvokeRequest,
+) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = handlePanic(Oneway, i.Options, r, i.Request.ToRequestMeta())
+		}
+	}()
+
+	return i.Handler.HandleOneway(i.Context, i.Request)
+}
+
+// InvokeStreamHandler calls the stream handler, recovering from panics as
+// errors.
+func InvokeStreamHandler(
+	i StreamInvokeRequest,
+) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = handlePanic(Streaming, i.Options, r, i.Stream.Request().Meta)
+		}
+	}()
+
+	return i.Handler.HandleStream(i.Stream)
+}
+
+func handlePanic(rpcType Type, options *InvokerOptions, recovered interface{}, req *RequestMeta) error {
+	err := fmt.Errorf("panic: %v", recovered)
+	if options != nil {
+		if options.logger != nil {
+			logPanic(rpcType, options.logger, err, req)
+			return err
+		}
+	}
+	log.Printf("%s handler panicked: %v\n%s", rpcType, recovered, debug.Stack())
+	return err
+}
+
+func logPanic(rpcType Type, logger *zap.Logger, err error, req *RequestMeta) {
+	logger.Error(fmt.Sprintf("%s handler panicked", rpcType),
+		zap.String("service", req.Service),
```

Comment:
I encountered this method (literally) today: https://godoc.org/go.uber.org/yarpc/api/transport#Request.MarshalLogObject
which _should_ take care of the field marshalling for you if you use `zap.Object()`.

Reply:
I figured out I can call https://godoc.org/go.uber.org/yarpc/api/transport#RequestMeta.ToRequest, but I want to confirm locally that the logging looks basically the same. I am concerned that it could change how the data gets formatted for sentry. I don't want to make that any worse.

Reply:
I confirmed locally that using the MarshalLogObject approach yields something like this:
```
Unary handler panicked  {"request": {"caller": "yab-ztrain", "service": "lottery", "encoding": "thrift", "procedure": "Lottery::doPanic", "shardKey": "", "routingKey": "", "routingDelegate": ""}, "error": "panic: PANIC: 9", "stack":
```
and I prefer the current format like this:
```
Unary handler panicked  {"service": "lottery", "transport": "tchannel", "procedure": "Lottery::doPanic", "encoding": "thrift", "caller": "yab-ztrain", "error": "panic: PANIC: 9", "stack":
```

Reply:
Ahh yes, my mistake, I didn't realize we were using a `RequestMeta` initially.
My primary concern here is that if we add more transport headers (eg #1520) , we'd always need to remember to add it here for parity.

We could add a `MarshalLogObject` method to the `RequestMeta` object as well.

Since this is primarily a vanity change for better logging, I'll leave this up to you! 👍